usr/src/man/man1/localedef.1
.\" Copyright (c) 1992, X/Open Company Limited  All Rights Reserved
.\" Portions Copyright (c) 2003, Sun Microsystems, Inc.  All Rights Reserved
.\" Portions Copyright 2013 DEY Storage Systems, Inc.
.\" Sun Microsystems, Inc. gratefully acknowledges The Open Group for
.\" permission to reproduce portions of its copyrighted documentation.
.\" Original documentation from The Open Group can be obtained online at
.\" http://www.opengroup.org/bookstore/.
.\" The Institute of Electrical and Electronics Engineers and The Open Group,
.\" have given us permission to reprint portions of their documentation. In
.\" the following statement, the phrase "this text" refers to portions of the
.\" system documentation. Portions of this text are reprinted and reproduced
.\" in electronic form in the Sun OS Reference Manual, from IEEE Std 1003.1,
.\" 2004 Edition, Standard for Information Technology -- Portable Operating
.\" System Interface (POSIX), The Open Group Base Specifications Issue 6,
.\" Copyright (C) 2001-2004 by the Institute of Electrical and Electronics
.\" Engineers, Inc and The Open Group. In the event of any discrepancy between
.\" these versions and the original IEEE and The Open Group Standard, the
.\" original IEEE and The Open Group Standard is the referee document. The
.\" original Standard can be obtained online at
.\" http://www.opengroup.org/unix/online.html.
.\"  This notice shall appear on any product containing this material.
.\" The contents of this file are subject to the terms of the Common
.\" Development and Distribution License (the "License").  You may not use
.\" this file except in compliance with the License.
.\" You can obtain a copy of the license at usr/src/OPENSOLARIS.LICENSE or
.\" http://www.opensolaris.org/os/licensing.  See the License for the specific
.\" language governing permissions and limitations under the License.
.\" When distributing Covered Code, include this CDDL HEADER in each file and
.\" include the License file at usr/src/OPENSOLARIS.LICENSE.  If applicable,
.\" add the following below this CDDL HEADER, with the fields enclosed by
.\" brackets "[]" replaced with your own identifying information:
.\" Portions Copyright [yyyy] [name of copyright owner]
.\"
.\" $FreeBSD: head/usr.bin/localedef/localedef.1 290514 2015-11-07 21:07:40Z bapt $
.\"
.Dd November 14, 2016
.Dt LOCALEDEF 1
.Os
.Sh NAME
.Nm localedef
.Nd define locale environment
.Sh SYNOPSIS
.Nm
.Op Fl cDUv
.Op Fl f Ar charmap
.Op Fl i Ar sourcefile
.Op Fl u Ar codeset
.Op Fl w Ar widthfile
.Ar localename
.Sh DESCRIPTION
The
.Nm
utility converts source definitions for locale categories
into a format usable by the functions and utilities whose operational behavior
is determined by the setting of the locale environment variables; see
.Xr environ 7 .
.Pp
The utility reads source definitions for one or more locale categories
belonging to the same locale from the file named in the
.Fl i
option (if specified) or from standard input.
.Pp
Each category source definition is identified by the corresponding environment
variable name and terminated by an
.Sy END
.Em category-name
statement.
The following categories are supported:
.Bl -tag -width ".Ev LC_MONETARY"
.It Ev LC_CTYPE
Defines character classification and case conversion.
.It Ev LC_COLLATE
Defines collation rules.
.It Ev LC_MONETARY
Defines the format and symbols used in formatting of monetary information.
.It Ev LC_NUMERIC
Defines the decimal delimiter, grouping and grouping symbol for non-monetary
numeric editing.
.It Ev LC_TIME
Defines the format and content of date and time information.
.It Ev LC_MESSAGES
Defines the format and values of affirmative and negative responses.
.El
.Pp
The following options are supported:
.Bl -tag -width indent
.It Fl c
Creates permanent output even if warning messages have been issued.
.It Fl D
BSD-style
output.
Rather than the default of creating the
.Ar localename
directory and creating files like
.Pa LC_CTYPE ,
.Pa LC_COLLATE ,
etc.\& in that directory,
the output file names have the format
.Dq <localename>.<category>
and are dumped to the current directory.
.It Fl f Ar charmap
Specifies the pathname of a file containing a mapping of character symbols and
collating element symbols to actual character encodings.
This option must be specified if symbolic names (other than collating symbols
defined in a
.Sy collating-symbol
keyword) are used.
If the
.Fl f
option is not present, the default character mapping will be used.
.It Fl i Ar sourcefile
The path name of a file containing the source definitions.
If this option is not present, source definitions will be read from
standard input.
.It Fl u Ar codeset
Specifies the name of a codeset used as the target mapping of character symbols
and collating element symbols whose encoding values are defined in terms of the
ISO/IEC 10646-1:2000 standard position constant values.
See
.Sx NOTES .
.It Fl U
Ignore the presence of character symbols that have no matching character
definition.  This facilitates the use of a common locale definition file
to be used across multiple encodings, even when some symbols are not
present in a given encoding.
.It Fl v
Emit verbose debugging output on standard output.
.It Fl w Ar widthfile
The path name of the file containing character screen width definitions.
If not supplied, then default screen widths will be assumed, which will
generally not account for East Asian encodings requiring more than a single
character cell to display, nor for combining or accent marks that occupy
no additional screen width.
.El
.Pp
The following operands are required:
.Bl -tag -width ".Ar localename"
.It Ar localename
Identifies the locale.
If the name contains one or more slash characters,
.Ar localename
will be interpreted as a path name where the created locale
definitions will be stored.
This capability may be restricted to users with appropriate privileges.
(As a consequence of specifying one
.Ar localename ,
although several categories can be processed in one execution, only categories
belonging to the same locale can be processed.)
.El
.Sh OUTPUT
.Nm
creates a directory of files that represents the locale's data,
unless instructed otherwise by the
.Fl D (
BSD
output) option.
The contants of this directory should generally be copied into the
appropriate subdirectory of
.Pa /usr/share/locale
in order the definitions to be visible to programs linked with libc.
.Sh ENVIRONMENT
See
.Xr environ 7
for definitions of the following environment variables that affect the
execution of
.Nm :
.Ev LANG ,
.Ev LC_ALL ,
.Ev LC_COLLATE ,
.Ev LC_CTYPE ,
.Ev LC_MESSAGES ,
.Ev LC_MONETARY ,
.Ev LC_MUMERIC ,
.Ev LC_TIME ,
and
.Ev NLSPATH .
.Sh EXIT STATUS
The following exit values are returned:
.Bl -tag -width XX
.It 0
No errors occurred and the locales were successfully created.
.It 1
Warnings occurred and the locales were successfully created.
.It 2
The locale specification exceeded implementation limits or the coded character
set or sets used were not supported by the implementation, and no locale was
created.
.It >3
Warnings or errors occurred and no output was created.
.El
.Pp
If an error is detected, no permanent output will be created.
.Sh SEE ALSO
.Xr locale 1 ,
.Xr iconv_open 3 ,
.Xr nl_langinfo 3 ,
.Xr strftime 3 ,
.Xr environ 7
.Sh WARNINGS
If warnings occur, permanent output will be created if the
.Fl c
option was specified.
The following conditions will cause warning messages to be issued:
.Bl -bullet
.It
If a symbolic name not found in the
.Pa charmap
file is used for the descriptions of the
.Sy LC_CTYPE
or
.Sy LC_COLLATE
categories (for other categories, this will be an error condition).
.It
If optional keywords not supported by the implementation are present in the
source.
.El
.Sh NOTES
When the
.Fl u
option is used, the
.Ar codeset
option-argument is interpreted as a name of a codeset to which the
ISO/IEC 10646-1:2000 standard position constant values are converted.
Both the ISO/IEC 10646-1:2000 standard position constant values and
other formats (decimal, hexadecimal, or octal) are valid as encoding
values within the charmap file.
The
.Ar codeset
can be any codeset that is supported by the
.Fn iconv_open 3
function.
.Pp
When conflicts occur between the charmap specification of
.Ar codeset ,
.Em mb_cur_max ,
or
.Em mb_cur_min
and the corresponding value for the codeset represented by the
.Fl u
option-argument
.Ar codeset ,
the
.Nm
utility fails with an error.
.Pp
When conflicts occur between the charmap encoding values specified for symbolic
names of characters of the portable character set and the character encoding
values defined by the US-ASCII, the result is unspecified.
.Sh HISTORY
.Nm
was written by
.An Garrett D'Amore
.Aq Mt garrett@nexenta.com
for Illumos.
.An John Marino
.Aq Mt draco@marino.st
provided the alternations necessary to compile cleanly on
.Dx .
.An Baptiste Daroussin
.Aq Mt bapt@FreeBSD.org
ported it to
.Fx
and converted it to
.Xr tree 3 .


usr/src/cmd/localedef/localedef.c
/*
 * Copyright 2010 Nexenta Systems, Inc.  All rights reserved.
 * Copyright 2013 DEY Storage Systems, Inc.
 * Copyright 2015 John Marino <draco@marino.st>
 *
 * This source code is derived from the illumos localedef command, and
 * provided under BSD-style license terms by Nexenta Systems, Inc.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
 * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
 * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
 * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
 * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
 * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * POSIX localedef.
 */

#include <stdio.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <string.h>
#include <unistd.h>
#include <libgen.h>
#include <stddef.h>
#include <unistd.h>
#include <limits.h>
#include <locale.h>
#include <dirent.h>
#include "localedef.h"
#include "parser.tab.h"

#ifndef	TEXT_DOMAIN
#define	TEXT_DOMAIN	"SYS_TEST"
#endif

static int bsd = 0;
int verbose = 0;
int undefok = 0;
int warnok = 0;
static char *locname = NULL;
static char locpath[PATH_MAX];

const char *
category_name(void)
{
	switch (get_category()) {
	case T_CHARMAP:
		return ("CHARMAP");
	case T_WIDTH:
		return ("WIDTH");
	case T_COLLATE:
		return ("LC_COLLATE");
	case T_CTYPE:
		return ("LC_CTYPE");
	case T_MESSAGES:
		return ("LC_MESSAGES");
	case T_MONETARY:
		return ("LC_MONETARY");
	case T_NUMERIC:
		return ("LC_NUMERIC");
	case T_TIME:
		return ("LC_TIME");
	default:
		INTERR;
		return (NULL);
	}
}

static char *
category_file(void)
{
	if (bsd)
		(void) snprintf(locpath, sizeof (locpath), "%s.%s",
		    locname, category_name());
	else
		(void) snprintf(locpath, sizeof (locpath), "%s/%s/LCL_DATA",
		    locname, category_name());
	return (locpath);
}

FILE *
open_category(void)
{
	FILE *file;

	if (verbose) {
		(void) printf(_("Writing category %s: "), category_name());
		(void) fflush(stdout);
	}

	/* make the parent directory */
	if (!bsd)
		(void) mkdirp(dirname(category_file()), 0755);

	/*
	 * note that we have to regenerate the file name, as dirname
	 * clobbered it.
	 */
	file = fopen(category_file(), "w");
	if (file == NULL) {
		errf(strerror(errno));
		return (NULL);
	}
	return (file);
}

void
close_category(FILE *f)
{
	if (fchmod(fileno(f), 0644) < 0) {
		(void) fclose(f);
		(void) unlink(category_file());
		errf(strerror(errno));
	}
	if (fclose(f) < 0) {
		(void) unlink(category_file());
		errf(strerror(errno));
	}
	if (verbose) {
		(void) fprintf(stdout, _("done.\n"));
		(void) fflush(stdout);
	}
}

/*
 * This function is used when copying the category from another
 * locale.  Note that the copy is actually performed using a hard
 * link for efficiency.
 */
void
copy_category(char *src)
{
	char	srcpath[PATH_MAX];
	int	rv;

	(void) snprintf(srcpath, sizeof (srcpath), "%s/%s/LCL_DATA",
	    src, category_name());
	rv = access(srcpath, R_OK);
	if ((rv != 0) && (strchr(srcpath, '/') == NULL)) {
		/* Maybe we should try the system locale */
		(void) snprintf(srcpath, sizeof (srcpath),
		    "/usr/lib/locale/%s/%s/LCL_DATA", src, category_name());
		rv = access(srcpath, R_OK);
	}

	if (rv != 0) {
		errf(_("source locale data unavailable"), src);
		return;
	}

	if (verbose > 1) {
		(void) printf(_("Copying category %s from %s: "),
		    category_name(), src);
		(void) fflush(stdout);
	}

	/* make the parent directory */
	if (!bsd)
		(void) mkdirp(dirname(category_file()), 0755);

	if (link(srcpath, category_file()) != 0) {
		errf(_("unable to copy locale data: %s"), strerror(errno));
		return;
	}
	if (verbose > 1) {
		(void) printf(_("done.\n"));
	}
}

int
putl_category(const char *s, FILE *f)
{
	if (s && fputs(s, f) == EOF) {
		(void) fclose(f);
		(void) unlink(category_file());
		errf(strerror(errno));
		return (EOF);
	}
	if (fputc('\n', f) == EOF) {
		(void) fclose(f);
		(void) unlink(category_file());
		errf(strerror(errno));
		return (EOF);
	}
	return (0);
}

int
wr_category(void *buf, size_t sz, FILE *f)
{
	if (!sz) {
		return (0);
	}
	if (fwrite(buf, sz, 1, f) < 1) {
		(void) fclose(f);
		(void) unlink(category_file());
		errf(strerror(errno));
		return (EOF);
	}
	return (0);
}

int yyparse(void);

static void
usage(void)
{
	(void) fprintf(stderr,
	    _("Usage: localedef [options] localename\n"));
	(void) fprintf(stderr, ("[options] are:\n"));
	(void) fprintf(stderr, ("  -c          : ignore warnings\n"));
	(void) fprintf(stderr, ("  -D          : BSD-style output\n"));
	(void) fprintf(stderr, ("  -v          : verbose output\n"));
	(void) fprintf(stderr, ("  -U          : ignore undefined symbols\n"));
	(void) fprintf(stderr, ("  -f charmap  : use given charmap file\n"));
	(void) fprintf(stderr, ("  -u encoding : assume encoding\n"));
	(void) fprintf(stderr, ("  -w widths   : use screen widths file\n"));
	(void) fprintf(stderr, ("  -i locsrc   : source file for locale\n"));
	exit(4);
}

int
main(int argc, char **argv)
{
	int c;
	char *lfname = NULL;
	char *cfname = NULL;
	char *wfname = NULL;
	DIR *dir;

	init_charmap();
	init_collate();
	init_ctype();
	init_messages();
	init_monetary();
	init_numeric();
	init_time();

	yydebug = 0;

	(void) setlocale(LC_ALL, "");
	(void) textdomain(TEXT_DOMAIN);

	while ((c = getopt(argc, argv, "w:i:cf:u:vUD")) != -1) {
		switch (c) {
		case 'D':
			bsd = 1;
			break;
		case 'v':
			verbose++;
			break;
		case 'i':
			lfname = optarg;
			break;
		case 'u':
			set_wide_encoding(optarg);
			break;
		case 'f':
			cfname = optarg;
			break;
		case 'U':
			undefok++;
			break;
		case 'c':
			warnok++;
			break;
		case 'w':
			wfname = optarg;
			break;
		case '?':
			usage();
			break;
		}
	}

	if ((argc - 1) != (optind)) {
		usage();
	}
	locname = argv[argc - 1];
	if (verbose) {
		(void) printf(_("Processing locale %s.\n"), locname);
	}

	if (cfname) {
		if (verbose)
			(void) printf(_("Loading charmap %s.\n"), cfname);
		reset_scanner(cfname);
		(void) yyparse();
	}

	if (wfname) {
		if (verbose)
			(void) printf(_("Loading widths %s.\n"), wfname);
		reset_scanner(wfname);
		(void) yyparse();
	}

	if (verbose) {
		(void) printf(_("Loading POSIX portable characters.\n"));
	}
	add_charmap_posix();

	if (lfname) {
		reset_scanner(lfname);
	} else {
		reset_scanner(NULL);
	}

	/* make the directory for the locale if not already present */
	if (!bsd) {
		while ((dir = opendir(locname)) == NULL) {
			if ((errno != ENOENT) ||
			    (mkdir(locname, 0755) <  0)) {
				errf(strerror(errno));
			}
		}
		(void) closedir(dir);
		(void) mkdirp(dirname(category_file()), 0755);
	}

	(void) yyparse();
	if (verbose) {
		(void) printf(_("All done.\n"));
	}
	return (warnings ? 1 : 0);
}
